README.md
TermLog v0.2.0
==============

Simple level based terminal logger


Features
--------

* Conditional Level Logging: uses a different level depending if the condition is `true` or `false`
* Custom datetime formatting
* Custom Fatal level exit code
* Custom Panic level exit code
* Forcing ANSI color on or off for level labels
* Logger namespacing
* Outputs to `stderr` by default but allows assignment of any `io.Writter`


Usage Examples
--------------

```go
package main

import "github.com/runeimp/termlog"

func main() {
	log := termlog.New()

	log.Debug("Debugging message")
	log.Info("Informational message")
	log.Warn("Warning message")
	log.Error("Error message")
}
```


### Conditional Level Logging

```go
package main

import "github.com/runeimp/termlog"

func main() {
	value := 0

	log := termlog.New()
	conditionTrueLevel := termlog.WarnLevel
	conditionFalseLevel := termlog.DebugLevel
	log.ConditionalLevel(value == 0, conditionTrueLevel, conditionFalseLevel, "If the value equals zero pay attention!")
	// Output: 2023-02-06 19:20:14.192617 WARN  If the value equals zero pay attention!
}
```

### If You Wish to Control All Things and Use a Namespace

```go
package main

import "github.com/runeimp/termlog"

func main() {
	log := termlog.New("my-namespace")
	log.ForceColor = termlog.ForceColorOff // Force off ANSI color rendering on supported systems
	log.Level = termlog.WarnLevel          // TermLog defaults to DebugLevel
	log.TimeFormat = "2006-01-02 15:04:05" // Default format but with no fractional seconds
	log.FatalExitCode = 13
	log.PanicExitCode = 42

	log.Debug("Debugging message")
	log.Info("Informational message")
	log.Warn("Warning message")
	log.Error("Error message")
	log.Fatal("Fatal message")
	log.Panic("Panic message")
}
```


### You Can Force Color if You Desire

If You Run Windows and don't use MS Terminal you can force TermLog to output ANSI color codes. If your system already has color support turned on it will simply work. If not you'll see the ANSI codes as text. ANSI color support should be active by default for later versions of Windows 10, all versions of Windows 11 and on. I just haven't built in testing for that support yet. So you have to force it to see if it's supported or not and to use it outside of MS Terminal currently.

```go
package main

import "github.com/runeimp/termlog"

func main() {
	log := termlog.New()
	log.ForceColor = termlog.ForceColorOn

	log.Debug("Debugging message")
	log.Info("Informational message")
	log.Warn("Warning message")
	log.Error("Error message")
	log.Fatal("Fatal message")
	log.Panic("Panic message")
}
```


Example Output
--------------

```sh
$ tester; echo "Exit Code: $?"
==> Testing Logger Up to Error Level With Info Level Set and ForceColor is ForceColorOff

2023-02-06 19:20:14.192486 INFO  Informational message
2023-02-06 19:20:14.192594 WARN  Warning message
2023-02-06 19:20:14.192596 ERROR Error message

==> Testing Namespaced Logger with Custom TimeFormat Up to Error Level With Default Settings

2023-02-06 19:20:14 my-namespace DEBUG Debugging message
2023-02-06 19:20:14 my-namespace INFO  Informational message
2023-02-06 19:20:14 my-namespace WARN  Warning message
2023-02-06 19:20:14 my-namespace ERROR Error message

==> Testing Logger with the ConditionalLevel method

2023-02-06 19:20:14.192609 INFO  This is condition: true  # Should be Info Log Level in this test
2023-02-06 19:20:14.192617 WARN  This is condition: false # Should be Warn Log Level in this test

==> Testing Logger Up to Panic Level With Custom Exit Codes and ForceColor is ForceColorOn

2023-02-06 19:20:14.192621 DEBUG Debugging message
2023-02-06 19:20:14.192623 INFO  Informational message
2023-02-06 19:20:14.192624 WARN  Warning message
2023-02-06 19:20:14.192626 ERROR Error message
2023-02-06 19:20:14.192627 FATAL Fatal message
exit status 13
```

Above represents the output when the library is used on an unsupported system i.e.: Windows without using MS Terminal or MSYS (Git Bash, MinGW, MinGW-64, etc.). On supported systems (everything else I've tested thus far) the level colors <span style="background: black; color: white; padding: 0.2em">(<span style="color:silver; font-weight:bold;">DEBUG</span>, <span style="color:cyan; font-weight:bold;">INFO</span>, <span style="color:yellow; font-weight:bold;">WARN</span>, <span style="color:red; font-weight:bold;">ERROR</span>, <span style="color:red; font-weight:bold;">FATAL</span>, and <span style="color:red; font-weight:bold;">PANIC</span>)</span> are displayed on non-Windows systems and on Windows when using MS Terminal or ANSI colors are forced on. I may eventually add testing for color support for PowerShell, etc. on Windows when used outside of MS Terminal as well. For now you can test this with `ForceColor` as noted above in the Usage Examples.




[MinGW 32/64 enable ANSI color sequences on Windows 10]: https://gist.github.com/fleroviux/8343879d95a72140274535dc207f467d



CHANGELOG.md
ChangeLog
=========

Version 0.2.0
-------------

2023-02-06

* Added ability to change the output target from os.Stderr to any io.Writer
* Added Logger.ConditionalLevel() method which outputs at a different logging level based on a boolean condition
* Export the LoggingLevel type
* Updated tests to cover new features


Version 0.1.0
-------------

2022-08-20

* Initial release


termlog_test.go
package termlog

import (
	"bytes"
	"fmt"
	"testing"
	"time"
)

const (
	testingDateFormat     = `2006-01-02`
	testingDateTimeFormat = `2006-01-02 15:04:05`
)

// TestTermLogConditionalLevelFalse type test
func TestTermLogConditionalLevelFalse(t *testing.T) {
	var (
		condition  = false
		output     bytes.Buffer
		want       = ""
		wantMsg    = "This is conditional: false"
		wantFormat = "%s \x1b[1;36mINFO \x1b[0m %s\n"
	) // "2023-02-06 \x1b[1;33mWARN \x1b[0m Unit Testing Output\n"

	want = fmt.Sprintf(wantFormat, time.Now().Format(testingDateFormat), wantMsg)

	logger := New()
	logger.Output = &output // Redirecting the output from os.Stderr for testing
	logger.TimeFormat = testingDateFormat

	logger.ConditionalLevel(condition, WarnLevel, InfoLevel, wantMsg)

	got := output.String()
	if got != want {
		t.Fatalf(`got = %q, want match for %q`, got, want)
	}
}

// TestTermLogConditionalLevelTrue type test
func TestTermLogConditionalLevelTrue(t *testing.T) {
	var (
		condition  = true
		output     bytes.Buffer
		want       = ""
		wantMsg    = "This is conditional: true"
		wantFormat = "%s \x1b[1;33mWARN \x1b[0m %s\n"
	) // "2023-02-06 \x1b[1;33mWARN \x1b[0m Unit Testing Output\n"

	want = fmt.Sprintf(wantFormat, time.Now().Format(testingDateFormat), wantMsg)

	logger := New()
	logger.Output = &output // Redirecting the output from os.Stderr for testing
	logger.TimeFormat = testingDateFormat

	logger.ConditionalLevel(condition, WarnLevel, InfoLevel, wantMsg)

	got := output.String()
	if got != want {
		t.Fatalf(`got = %q, want match for %q`, got, want)
	}
}

// TestTermLogOutput test
func TestTermLogOutput(t *testing.T) {
	var (
		output     bytes.Buffer
		want       = ""
		wantFormat = "%s \x1b[1;33mWARN \x1b[0m %s\n"
		wantMsg    = "Unit Testing Output"
	)

	want = fmt.Sprintf(wantFormat, time.Now().Format(testingDateFormat), wantMsg)

	logger := New()
	logger.Output = &output // Redirecting the output from os.Stderr for testing
	logger.TimeFormat = testingDateFormat
	logger.Warn(wantMsg)

	got := output.String()
	if got != want {
		t.Fatalf(`got = %q, want match for %q`, got, want)
	}
}

// TestTermLogStringDefault string test
func TestTermLogStringDefault(t *testing.T) {
	logger := New()
	got := fmt.Sprintf("%s", logger)
	want := `*termlog.Logger{level: 5, namespace: ""}`
	if got != want {
		t.Fatalf(`got = %q, want match for %q`, got, want)
	}
}

// TestTermLogStringNamespaced string test
func TestTermLogStringNamespaced(t *testing.T) {
	logger := New("myname")
	got := fmt.Sprintf("%s", logger)
	want := `*termlog.Logger{level: 5, namespace: "myname"}`
	if got != want {
		t.Fatalf(`got = %q, want match for %q`, got, want)
	}
}

// TestTermLogType type test
func TestTermLogType(t *testing.T) {
	logger := New()
	got := fmt.Sprintf("%T", logger)
	want := "*termlog.Logger"
	if got != want {
		t.Fatalf(`got = %q, want match for %q`, got, want)
	}
}


termlog.go
package termlog

import (
	"fmt"
	"io"
	"os"
	"runtime"
	"runtime/debug"
	"time"
)

const (
	defaultFormat     = `2006-01-02 15:04:05.000000` // Pythonesque Timestamp instead of 2006-01-02 15:04:05 Z0700
	SystemUnsupported = "Unsupported"
	Version           = "0.2.0"
)

const (
	DebugLevel LoggingLevel = 5 // Debugging messages
	ErrorLevel LoggingLevel = 2 // Error messages (default)
	FatalLevel LoggingLevel = 1 // Fatal message with exit code 1
	InfoLevel  LoggingLevel = 4 // Informational messages
	PanicLevel LoggingLevel = 0 // Panic message with exit code 1 and stack trace
	WarnLevel  LoggingLevel = 3 // Warning messages
)

const (
	FormatDebugColor       = "\033[0;37m%s\033[0m" // 0;37 White-Normal (Silver)
	FormatErrorColor       = "\033[1;31m%s\033[0m" // 1;31 Red-Bold
	FormatFatalColor       = "\033[1;31m%s\033[0m" // 1;31 Red-Bold
	FormatInformationColor = "\033[1;36m%s\033[0m" // 1;36 Cyan-Bold
	FormatPanicColor       = "\033[1;31m%s\033[0m" // 1;31 Red-Bold
	FormatWarningColor     = "\033[1;33m%s\033[0m" // 1;33 Yellow-Bold
)

const (
	ForceColorUndefined colorForce = "undefined"
	ForceColorOff       colorForce = "off"
	ForceColorOn        colorForce = "on"
)

const (
	ANSIPrefix             = "\033["   // Not yet utilized
	ANSIColorSuffix        = "m"       // Not yet utilized
	ANSIReset              = "\033[0m" // Not yet utilized
	DebugLabel             = "DEBUG"
	DebugLabelANSI         = "\033[0;37mDEBUG\033[0m"    // 0;37 White-Normal (Silver)
	DebugLabelBgANSI       = "\033[0;30;47mDEBUG\033[0m" // 0;30;47 FG: Black, BG: White-Normal (Silver)
	InformationLabel       = "INFO "
	InformationLabelANSI   = "\033[1;36mINFO \033[0m"    // 1;36 Cyan-Bold
	InformationLabelBgANSI = "\033[1;30;46mINFO \033[0m" // 1;30;46 FG: Black, BG: Cyan-Bold
	WarningLabel           = "WARN "
	WarningLabelANSI       = "\033[1;33mWARN \033[0m"     // 1;33 Yellow-Bold
	WarningLabelBgANSI     = "\033[1;30;103mWARN \033[0m" // 1;30;103 FG: Black, BG: Yellow-Bold
	ErrorLabel             = "ERROR"
	ErrorLabelANSI         = "\033[1;31mERROR\033[0m"     // 1;31 Red-Bold
	ErrorLabelBgANSI       = "\033[1;30;101mERROR\033[0m" // 1;30;101 FG: Black, BG: Red-Bold
	FatalLabel             = "FATAL"
	FatalLabelANSI         = "\033[1;31mFATAL\033[0m"     // 1;31 Red-Bold
	FatalLabelBgANSI       = "\033[1;30;101mFATAL\033[0m" // 1;30;101 FG: Black, BG: Red-Bold
	PanicLabel             = "PANIC"
	PanicLabelANSI         = "\033[1;31mPANIC\033[0m"     // 1;31 Red-Bold
	PanicLabelBgANSI       = "\033[1;30;101mPANIC\033[0m" // 1;30;101 FG: Black, BG: Red-Bold
)

// colorForce used to define a trinary value type
type colorForce string

// LoggingLevel is used to define logging levels
type LoggingLevel uint8

// Logger is the base structure for logging
type Logger struct {
	ForceColor    colorForce   // Force color output
	Level         LoggingLevel // The current logging level for this instance
	Namespace     string       // The namespace defined for this instance
	system        string       // The active operating system
	FatalExitCode int          // The exit code to use on Fatal messages
	Output        io.Writer    // The output target to use (Defaults to os.Stderr when using the New function)
	PanicExitCode int          // The exit code to use on Panic messages
	TimeFormat    string       // The time format to use
}

// Debug outputs a debugging level message
func (l *Logger) ConditionalLevel(condition bool, cTrue, cFalse LoggingLevel, format string, msg ...any) {
	if condition {
		if l.Level >= cTrue {
			l.printLog(cTrue, format, msg...)
		}
	} else {
		if l.Level >= cFalse {
			l.printLog(cFalse, format, msg...)
		}
	}
}

// Debug outputs a debugging level message
func (l *Logger) Debug(format string, msg ...any) {
	if l.Level > InfoLevel {
		l.printLog(DebugLevel, format, msg...)
	}
}

// Error outputs a error level message
func (l *Logger) Error(format string, msg ...any) {
	if l.Level > FatalLevel {
		l.printLog(ErrorLevel, format, msg...)
	}
}

// Panic outputs a panic level message with exit code
func (l *Logger) Fatal(msg ...any) {
	var format string

	if len(msg) > 0 {
		format = msg[0].(string)
		if len(msg) > 1 {
			msg = msg[1:]
		} else {
			msg = []any{}
		}
	}

	l.printLog(FatalLevel, format, msg...)
	os.Exit(l.FatalExitCode)
}

func (l *Logger) getLabel(level LoggingLevel) (result string) {
	/*
		ANSI Color Off
			- If the system is supported and Logger.ForceColor is ForceColorOff
			- If the system is unsupported and Logger.ForceColor is not ForceColorOn
		ANSI Color On
			- If the system is supported and Logger.ForceColor is not ForceColorOff
			- If the system is unsupported but Logger.ForceColor is ForceColorOn
	*/

	enableANSI := false
	if l.ForceColor != ForceColorUndefined {
		if l.ForceColor == ForceColorOn {
			enableANSI = true
		}
	} else if l.system != SystemUnsupported {
		enableANSI = true
	}

	switch level {
	case DebugLevel:
		result = DebugLabel
		if enableANSI {
			result = DebugLabelANSI
		}
	case InfoLevel:
		result = InformationLabel
		if enableANSI {
			result = InformationLabelANSI
			// result = fmt.Sprintf(FormatInformationColor, InformationLabel)
		}
	case WarnLevel:
		result = WarningLabel
		if enableANSI {
			result = WarningLabelANSI
			// result = fmt.Sprintf(FormatWarningColor, WarningLabel)
		}
	case ErrorLevel:
		result = ErrorLabel
		if enableANSI {
			result = ErrorLabelANSI
			// result = fmt.Sprintf(FormatErrorColor, ErrorLabel)
		}
	case FatalLevel:
		result = FatalLabel
		if enableANSI {
			result = FatalLabelANSI
			// result = fmt.Sprintf(FormatFatalColor, FatalLabel)
		}
	case PanicLevel:
		result = PanicLabel
		if enableANSI {
			result = PanicLabelANSI
		}
	}

	if len(l.Namespace) > 0 {
		result = fmt.Sprintf("%s %s", l.Namespace, result)
	}

	return result
}

// Info outputs a information level message
func (l *Logger) Info(format string, msg ...any) {
	if l.Level > WarnLevel {
		// printLog(InfoLevel, l.system, l.Namespace, format, msg...)
		l.printLog(InfoLevel, format, msg...)
	}
}

// Panic outputs a panic level message with exit code and stack trace
func (l *Logger) Panic(msg ...any) {
	var format string

	if len(msg) > 0 {
		format = msg[0].(string)
		if len(msg) > 1 {
			msg = msg[1:]
		} else {
			msg = []any{}
		}
	}

	// log.Panic(format)
	l.printLog(PanicLevel, format, msg...)
	fmt.Fprintln(os.Stderr)
	debug.PrintStack()
	// fmt.Fprintln(os.Stderr)
	os.Exit(l.PanicExitCode)
}

func (l *Logger) printLog(level LoggingLevel, format string, msg ...any) {
	fmt.Fprintf(l.Output, "%s %s %s\n", time.Now().Format(l.TimeFormat), l.getLabel(level), fmt.Sprintf(format, msg...))
}

func (l *Logger) String() string {
	return fmt.Sprintf("%T{level: %d, namespace: %q}", l, l.Level, l.Namespace)
}

// Warn outputs a warning level message
func (l *Logger) Warn(format string, msg ...any) {
	if l.Level > ErrorLevel {
		l.printLog(WarnLevel, format, msg...)
	}
}

// New returns a Logger
func New(args ...string) *Logger {
	var namespace string
	if len(args) > 0 {
		namespace = args[0]
	}

	isMSTerminalSession := false
	isMSYS := false
	system := runtime.GOOS

	if os.Getenv("WT_SESSION") != "" {
		isMSTerminalSession = true
	}
	if os.Getenv("MSYSTEM") != "" {
		isMSYS = true
	}
	// fmt.Printf("termlog.New() | Windows System: %q\n", system)
	/* NOTE: Test for unsupported systems goes here
	It works fine in MS Terminal, Terminal/PowerShell 5 Desktop, Terminal/Command Prompt,
	and PowerShell 7.2.5 Core on macOS Big Sur.
	*/
	if system == "windows" {
		// Test for Windows Terminal as it has ANSI support turned on by default
		if isMSTerminalSession != true && isMSYS != true { // If Windows Terminal is not hosting the session
			system = SystemUnsupported
		}
		// fmt.Printf("termlog.New() | Windows System: %q\n", system)
	}

	// fmt.Printf("termlog.New() | isMSTerminalSession: %t\n", isMSTerminalSession)
	// fmt.Printf("termlog.New() | shellOut: %s\n", shellOut)
	// if shellErr != nil {
	// 	fmt.Printf("termlog.New() | shellErr: %s\n", shellErr.Error())
	// }

	return &Logger{
		ForceColor:    ForceColorUndefined,
		Level:         DebugLevel,
		Namespace:     namespace,
		system:        system,
		FatalExitCode: 1,
		Output:        os.Stderr,
		PanicExitCode: 2,
		TimeFormat:    defaultFormat,
	}
}


cmd/tester/tester.go
package main

import (
	"fmt"

	"github.com/runeimp/termlog"
)

func main() {
	fmt.Printf("==> Testing Logger Up to Error Level With Info Level Set and ForceColor is ForceColorOff\n\n")
	testInfoLogger()
	fmt.Printf("\n==> Testing Namespaced Logger with Custom TimeFormat Up to Error Level With Default Settings\n\n")
	testNamespacedLogger()
	fmt.Printf("\n==> Testing Logger with the ConditionalLevel method\n\n")
	testConditionalLevelLogger()
	fmt.Printf("\n==> Testing Logger Up to Panic Level With Custom Exit Codes and ForceColor is ForceColorOn\n\n")
	testCustomExitCodesLogger()
	fmt.Println()
}

func testInfoLogger() {
	log := termlog.New()
	log.ForceColor = termlog.ForceColorOff
	log.Level = termlog.InfoLevel
	log.Debug("Debugging message")
	log.Info("Informational message")
	log.Warn("Warning message")
	log.Error("Error message")
}

func testCustomExitCodesLogger() {
	log := termlog.New()
	log.ForceColor = termlog.ForceColorOn
	log.FatalExitCode = 13
	log.PanicExitCode = 42
	log.Debug("Debugging message")
	log.Info("Informational message")
	log.Warn("Warning message")
	log.Error("Error message")
	log.Fatal("Fatal message") // The default exit code 1
	log.Panic("Panic message") // The default exit code 2 preceded by a stack trace
}

func testNamespacedLogger() {
	log := termlog.New("my-namespace")
	log.TimeFormat = "2006-01-02 15:04:05"
	log.Debug("Debugging message")
	log.Info("Informational message")
	log.Warn("Warning message")
	log.Error("Error message")
	// log.Fatal("Fatal message") // Message then default exit code 1
	// log.Panic("Panic message") // Message then default exit code 2 and stack trace
}

func testConditionalLevelLogger() {
	log := termlog.New()
	log.Level = termlog.WarnLevel
	condition := 0 < 1
	log.ConditionalLevel(condition, termlog.InfoLevel, termlog.WarnLevel, "This is condition: %-5t # Should be Info Log Level in this test", condition)
	condition = false
	log.ConditionalLevel(condition, termlog.InfoLevel, termlog.WarnLevel, "This is condition: %5t # Should be Warn Log Level in this test", condition)
}
